### PR TITLE
HPCC-30776 ECL mods to support pluggable file format syntax *** WIP ***

### PR DESCRIPTION
**Pull request by dcamper**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
Looks like you're getting to grips with the structure.  I would also add a new operator no_filetype, create that in setPluggableModeExpr
e.g.

targetAttr.setExpr(createValue(no_filetype, makeNullType(), createAttribute(fileFormat), *options));

Then search for no_csv in the sources in ecl/hql and add corresponding entries.  You could also do the same for ecl/hqlcpp, but I would probably leave that to a separate PR, which meant the new syntax generated the generic disk read activity.

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -8992,6 +8992,29 @@ bool HqlGram::convertAllToAttribute(attribute &atr)
...
+    // the grammar should have ensured that none of the legacy (built-in) file formats
+    // were activated, even with the new TYPE(<format>) syntax; if you get here,
+    // you should be only trying to load a plugin
+    IAtom * fileFormat = lower(mode.getId());
+    StringBuffer fileFormatStr(fileFormat->queryStr());
+    
+    // TODO: Look for a plugin filetype of name fileFormatStr
+    DBGLOG("HqlGram::setPluggableModeExpr processing file type %s", fileFormatStr.str());
+
+	// Following is a placeholder to make the parser happy -- note the hardwiring
```
[the remark this answers is not in the document]
Comment (dcamper) [reply ecl/hql/hqlgram2.cpp:9007]: Fixed.

File: ecl/hql/hqlgram.y [hunk 2]
```diff
@@ -10529,6 +10529,75 @@ mode
                             $$.setExpr(createValue(no_json, makeNullType(), args));
                         }
     | pipe
+    | TYPE '(' FLAT ')'
```
[the remark this answers is not in the document]
Comment (dcamper) [reply ecl/hql/hqlgram.y:10532]: Also: A plugin-style filetype will process options differently than flex. This is definitely not a problem with a new filetype such as Parquet, but may cause backwards-compatibility issues with an existing filetype (errors may be caught at different times and/or reported differently). Is that a concern, given that this is introducing a new syntax?

**Comment by dcamper** [comment]
Closing while continuing work.  Will open a new draft later.

**Pull request closed without merging**